### PR TITLE
feat: enable POST requests batching with dl.

### PR DESCRIPTION
**Pull request by laststylebender14**
**Summary:**  
docs: https://github.com/tailcallhq/tailcallhq.github.io/pull/559

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [ ] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: src/core/blueprint/operators/http.rs [hunk 1]
```diff
@@ -16,8 +17,8 @@ pub fn compile_http(
 ) -> Valid<IR, String> {
     let dedupe = http.dedupe.unwrap_or_default();
 
-    Valid::<(), String>::fail("GroupBy is only supported for GET requests".to_string())
-        .when(|| !http.batch_key.is_empty() && http.method != Method::GET)
+    Valid::<(), String>::fail("GroupBy is only supported for GET/POST requests".to_string())
+        .when(|| !http.batch_key.is_empty() && !matches!(http.method, Method::GET | Method::POST))
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:21]: No need to check for methods at all.

File: src/core/blueprint/operators/http.rs [hunk 2]
```diff
@@ -112,3 +155,36 @@ pub fn update_http<'a>(
...
+/// template type.
+fn extract_expression_keys(json: &str) -> Vec<String> {
+    let mut keys = Vec::new();
+    let re = Regex::new(r#""([^"]+)"\s*:\s*"\{\{.*?\}\}""#).unwrap();
+    for cap in re.captures_iter(json) {
+        if let Some(key) = cap.get(1) {
+            keys.push(key.as_str().to_string());
+        }
+    }
+    println!("[Finder]: input: {:#?} and output: {:#?}", json, keys);
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:169]:
```suggestion
```

File: src/core/blueprint/operators/http.rs [hunk 3]
```diff
@@ -112,3 +155,36 @@ pub fn update_http<'a>(
         },
     )
 }
+
+/// extracts the keys from the json representation, if the value is of mustache
+/// template type.
+fn extract_expression_keys(json: &str) -> Vec<String> {
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:161]: - Make Body of Json Type
- Automatically get a dynamic value
- Iterate over the dynamic value and identify your key

File: src/core/blueprint/operators/http.rs [hunk 4]
```diff
@@ -63,13 +89,30 @@ pub fn compile_http(
...
+                            .then(|| q.key.clone())
+                    })
+                } else {
+                    // find the key from the body where value is mustache template.
+                    http.body
+                        .as_ref()
+                        .map(|b| extract_expression_keys(b))
+                        .and_then(|keys| {
+                            if keys.len() == 1 {
+                                Some(keys[0].clone())
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:109]: `keys.get(0)` is going to be safer and more concise.

File: src/core/blueprint/operators/http.rs [hunk 5]
```diff
@@ -63,13 +89,30 @@ pub fn compile_http(
...
+                            } else {
+                                None
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:111]: If you can't identify the key, you should throw an error.

File: src/core/blueprint/operators/http.rs [hunk 6]
```diff
@@ -28,6 +29,31 @@ pub fn compile_http(
...
             }),
         )
+        .and_then(|_| {
+            let result = if http.method == Method::POST {
+                if !http.batch_key.is_empty() {
+                    let keys = http
+                        .body
+                        .as_ref()
+                        .map(|b| extract_expression_keys(b).len())
+                        .unwrap_or_default();
```
Comment (tusharmath) [inline src/core/blueprint/operators/http.rs:39]: Move this check below where the key is actually used.

File: src/core/http/data_loader.rs [hunk 7]
```diff
@@ -71,12 +73,55 @@ impl Loader<DataLoaderRequest> for HttpDataLoader {
...
-            // TODO: enable in tests only
-            dl_requests.sort_by(|a, b| a.to_request().url().cmp(b.to_request().url()));
+            if cfg!(feature = "integration_test") || cfg!(test) {
+                dl_requests.sort_by(|a, b| a.to_request().url().cmp(b.to_request().url()));
+            }
 
             // Create base request
-            let mut request = dl_requests[0].to_request();
-            let first_url = request.url_mut();
+            let mut base_request = dl_requests[0].to_request();
```
Comment (tusharmath) [inline src/core/http/data_loader.rs:81]: Can fail for empty keys.

File: src/core/http/data_loader.rs [hunk 8]
```diff
@@ -71,12 +73,55 @@ impl Loader<DataLoaderRequest> for HttpDataLoader {
...
+            // TODO: add the body as is in the DalaLoaderRequest.
+            let mut request_to_body_map = HashMap::with_capacity(dl_requests.len());
+
+            if dl_requests[0].method() == reqwest::Method::POST {
+                // run only for POST requests.
+                let mut merged_body = Vec::with_capacity(dl_requests.len());
+                for req in dl_requests.iter() {
+                    if let Some(body) = req.body().and_then(|b| b.as_bytes()) {
```
Comment (tusharmath) [inline src/core/http/data_loader.rs:89]: deserializing request before making it is going to be expensive.

File: src/core/blueprint/operators/http.rs [hunk 9]
```diff
@@ -105,3 +127,78 @@ pub fn compile_http(
         })
         .and_then(apply_select)
 }
+
+/// extracts the keys from the json representation, if the value is of mustache
+/// template type.
+fn extract_expression_paths(json: &serde_json::Value) -> Vec<Vec<Cow<'_, str>>> {
```
Comment (laststylebender14) [inline src/core/blueprint/operators/http.rs:133]: optimise this flow

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -169,7 +169,7 @@ jobs:
       - name: Run Cargo Test
         if: matrix.test != 'false'
         # TODO: run llvm-cov only for single build since other builds are not sent to codecov anyway
-        run: cargo llvm-cov --workspace ${{ matrix.features }} --lcov --target ${{ matrix.target }} --output-path lcov.info
```
Comment (tusharmath) [inline .github/workflows/ci.yml:172]: Why did we change this?

Comment (laststylebender14) [reply .github/workflows/ci.yml:172]: we wanted to sort req urls/bodies but only for tests.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**